Explain k in approximate k-NN (#7194)
opensearch-project/documentation-website: _search-plugins/knn/approximate-knn.md

}
```

In the preceding query, `k` is the number of neighbors the search of each graph returns. You must also include the `size` option, indicating the final number of results you want the query to return.  

> For NMSLIB and Faiss engines, `k` is the maximum number of documents returned for all segments of a shard. For the Lucene engine, `k` is the number of documents returned for a shard. The maximum value of `k` is 10,000.

Reviewer: This line is not adding no new information if we don't add an example. An example like this:

| size | k  | Number of Shards | Number of segments per shard | Engine       | Final Number of Results Returned |
|------|----|------------------|------------------------------|--------------|----------------------------------|
| 10   | 1  | 1                | 4                            | Faiss/Nmslib | 4                                |
| 10   | 1  | 1                | 4                            | Lucene       | 1                                |
| 10   | 10 | 1                | 4                            | Faiss/Nmslib | 10                               |
| 10   | 10 | 1                | 4                            | Lucene       | 10                               |
| 10   | 1  | 2                | 4                            | Faiss/Nmslib | 8                                |
| 10   | 1  | 2                | 4                            | Lucene       | 2                                |

Author: @navneet1v Thank you for the additional information. I have added it to the docs. Please review again.

}
```

### The number of returned results

> In the preceding query, `k` is the number of neighbors the search of each graph returns. You must also include the `size` option, indicating the final number of results you want the query to return.  

Reviewer 2: ```suggestion
In the preceding query, `k` represents the number of neighbors returned by the search of each graph. You must also include the `size` option, indicating the final number of results that you want the query to return.  
```

> For NMSLIB and Faiss engines, `k` is the maximum number of documents returned for all segments of a shard. For the Lucene engine, `k` is the number of documents returned for a shard. The maximum value of `k` is 10,000.

Reviewer 2: ```suggestion
For NMSLIB and Faiss engines, `k` represents the maximum number of documents returned for all segments of a shard. For the Lucene engine, `k` represents the number of documents returned for a shard. The maximum value of `k` is 10,000.
```

> For NMSLIB and Faiss engines, `k` is the maximum number of documents returned for all segments of a shard. For the Lucene engine, `k` is the number of documents returned for a shard. The maximum value of `k` is 10,000.

Reviewer 2: Should "the" precede "NMSLIB and Faiss", as it does "Lucene"?

> For any engine, each shard returns `size` results to the coordinator node. Thus, the total number of results the coordinator node receives is `size * number of shards`. After the coordinator node consolidates the results received from all nodes, the query returns the top `size` results.

Reviewer 2: ```suggestion
For any engine, each shard returns `size` results to the coordinator node. Thus, the total number of results that the coordinator node receives is `size * number of shards`. After the coordinator node consolidates the results received from all nodes, the query returns the top `size` results.
```

> The following table provides examples of the number of results returned by various engines in several scenarios.

Reviewer: @kolchfa-aws why we are not adding the full table here?

Author: @navneet1v It is the full table, but each pair of two identical rows is one row that contains Faiss/NMSLIB and Lucene results side by side so it's easier for the users to compare those.

![Uploading Screenshot 2024-05-28 at 5.01.11 PM.png…]()

Reviewer: oh.. make sense. So you added 1 more column. thanks

Reviewer: Can we add 1 more line above this, saying that this the table assumes data there are enough number of documents across shards and segments.

Author: @navneet1v Done!

> The following table provides examples of the number of results returned by various engines in several scenarios. For these examples, assume that the number of documents in the segments and shards is sufficient to return the number of results specified in the table.

Reviewer 2: ```suggestion
The following table provides examples of the number of results returned by various engines in several scenarios. For these examples, assume that the number of documents contained in the segments and shards is sufficient to return the number of results specified in the table.
```

> `size` 	| `k` | Number of primary shards | 	Number of segments per shard | Number of returned results, Faiss/NMSLIB | Number of returned results, Lucene

Reviewer 2: "Number of returned Faiss/NMSLIB results"? "Number of returned Lucene results"?

Author: I think it's clearer if you put the engine last.

10 |	1 |	1 |	4 |	4 | 1
10 | 10 |	1 |	4 |	10 | 10
10 |	1 |	2 |	4 |	8 | 2
 

> Faiss/NMSLIB return the number of results different from Lucene only when `k` is smaller than `size`. If `k` and `size` are equal, all engines return the same number of results. 

Reviewer 2: ```suggestion
Faiss/NMSLIB return the number of results differently from Lucene only when `k` is smaller than `size`. If `k` and `size` are equal, all engines return the same number of results. 
```

> Starting with OpenSearch 2.14, you can use `k`, `min_score`, or `max_distance` for [radial search]({{site.url}}{{site.baseurl}}/search-plugins/knn/radial-search-knn/).

Reviewer 2: ```suggestion
Starting in OpenSearch 2.14, you can use `k`, `min_score`, or `max_distance` for [radial search]({{site.url}}{{site.baseurl}}/search-plugins/knn/radial-search-knn/).
```
